### PR TITLE
Minimize the set of multiple inheritance (coercion) paths to check for conversion

### PR DESCRIPTION
**Pull request by pi8027**
The coercion mechanism reports a pair of inheritance paths sharing the same source and target classes (hereinafter a diamond) as ambiguous paths if they are inconvertible, to let the user ensure the coherence of the inheritance graph. In general, we do not necessarily have to check all the diamonds in an inheritance graph to verify its coherence. For example, it is sufficient to check two (small) diamonds `([f1; g2], [g1, f2])` and `([f2; h2], [h1; f3])` in the following diagram, since commutations of other (large) diamonds can be deduced from those of small ones using basic equational reasoning. In practice, not checking such large diamonds may reduce the number of ambiguous path warnings.
```
     /\
    /  \
f1 /    \ g1
  /      \
 /        \
 \        /\
  \   f2 /  \
g2 \    /    \ h1
    \  /      \
     \/        \
      \        /
       \      /
     h2 \    / f3
         \  /
          \/
```
This PR implements the most general way to do such reduction I could find, which is [proved correct](https://gist.github.com/pi8027/aa63dd0f4f575788a9d0e3f927539e52#file-meta-v-L407) in the sense that the inheritance graph is coherent if Coq does not report any ambiguous path. (I also have a draft paper briefly explain this, which I think I can partially share if needed.) To check if a diamond is reducible or not, the table of coercion classes (`class_tab`) has been extended with:
1. the two sets of coercion classes reachable from (`cl_reachable_from`) and to (`cl_reachable_to`) the class, respectively, and
2. the representative element of the strongly connected component including the class (`cl_repr`).

(I will add more explanations later depending on what reviewers request. Let me run CI first.)

**Kind:** enhancement.

- [x] Added / updated test-suite

- ~Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).~
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by pi8027** [comment]
BTW, I still do not understand the following things in coercionops.ml:
- `Bijint` and `cl_index` are really needed? Replacing them with `ClTypMap` and `cl_typ` would work and simplify the implementation. (I believe there is no performance issue here. Another potential issue would be the printing order of `Print Graph`.)
- A circular inheritance path of `C >-> C` can be registered to `inheritance_graph` only if `different_class_params` holds for `C`. What are the use cases of this?

**Comment by gares** [comment]
>     * `Bijint` and `cl_index` are really needed? Replacing them with `ClTypMap` and `cl_typ` would work and simplify the implementation. (I believe there is no performance issue here. Another potential issue would be that the printing order of `Print Graph` will change.)

I'm OK with that. BTW, if you are going to refactor the code, in spite of #13902 being merged, I still don't have an API that given a GlobRef.t tells me the number of parameters *and* the src/tgt classes. Would you mind making it possible?

> A circular inheritance path of `C >-> C` can be registered to `inheritance_graph` only if `different_class_params` holds for `C`. What are the use cases of this?

If I'm not mistake these are not "loops" but rather a form of "identity coercions": indexing looking just at the head constant is too simplistic.

**Comment by pi8027** [comment]
> > `Bijint` and `cl_index` are really needed? Replacing them with `ClTypMap` and `cl_typ` would work and simplify the implementation. (I believe there is no performance issue here. Another potential issue would be that the printing order of `Print Graph` will change.)
> 
> I'm OK with that. BTW, if you are going to refactor the code, in spite of #13902 being merged, I still don't have an API that given a GlobRef.t tells me the number of parameters _and_ the src/tgt classes. Would you mind making it possible? (adding the missing fields to coe_info_typ should suffice)

Sure. I will do that first and then redo this PR on top of it.

> > A circular inheritance path of `C >-> C` can be registered to `inheritance_graph` only if `different_class_params` holds for `C`. What are the use cases of this?
> 
> If I'm not mistake these are not "loops" but rather a form of "identity coercions": indexing looking just at the head constant is too simplistic.

I don't think so. In the following condition, `different_class_params env i` is relevant only if the source class `i` is equal to the target class `j`. If it is an identity coercion, the source class should be a constant that unfolds to the target class applied to some arguments; thus, they are not the same, as you said.
https://github.com/coq/coq/blob/4a0cf0f3fb4a1a57627f22897030771b2921ce85/pretyping/coercionops.ml#L341

**Comment by gares** [comment]
I think you can nor rebase on top of the refactoring PR which was merged

**Comment by pi8027** [comment]
I don't think I have to update the reference manual in accordance with this change. So this PR is ready for review.

File: pretyping/coercionops.mli [hunk 1]
```diff
@@ -31,9 +31,19 @@ val subst_cl_typ : env -> substitution -> cl_typ -> cl_typ
...
-  cl_param : int }
+  (* The number of parameters of the coercion class. *)
+  cl_param : int;
+  (* The sets of coercion classes respectively reachable from and to the
+     coercion class. *)
+  cl_reachable_from : ClTypSet.t;
+  cl_reachable_to : ClTypSet.t;
+  (* The representative class of the strongly connected component. *)
+  cl_repr : cl_typ;
+}
```
Comment (SkySkimmer) [inline pretyping/coercionops.mli:46]: We don't need to expose all this stuff, just expose a function `class_nparams : cl_typ -> int` and remove `class_info`, `cl_info_typ` and `ClTypSet`

**Comment by gares** [comment]
@pi8027 can you take of the request by @SkySkimmer ? Otherwise this seems ready to me.

**Comment by pi8027** [comment]
Yes. I will do that by the end of this week.

**Comment by gares** [comment]
@coqbot merge now

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
@coqbot merge now

**Pull request merged**